### ISSUE TITLE
Jittery videos (AVC1)

### ISSUE THREAD

**Issue by Munzu** [elFarto/nvidia-vaapi-driver#43]
I use the Firefox plugin h264ify to force YouTube videos to use the AVC1 codec since my GPU doesn't support decoding VP9. 
I noticed that YouTube videos and Twitch streams are very jittery. It's particularly noticeable at 60fps. Audio is completely fine. 
It's hard to explain exactly and I don't know if this is the right term for it but it feels like the frame timings are wrong. Sometimes a frame is shown for too long, sometimes it jumps into the future by a frame or more. Most of the times this happens without any frames being dropped according to "Stats for nerds". I notice these jitters between multiple times a second to once every 5 seconds or so.
~~The h264ify plugin to my knowledge does not affect Twitch and when I disable it, I still notice the jitters on Twitch.~~
Edit: I can observe the jitters on some Twitch streams and VODS but not always.

Among others, I tested and noticed it with videos like these and I think they highlight jitters quite clearly due to the smooth nature of the footage:
https://www.youtube.com/playlist?list=UUlcA5er9C4GQjOf-oLyZXIA

- Arch
- Kernel 5.16.2-arch1-1
- GTX 970
- Proprietary driver version 495.46
- X11
- KDE Plasma
- Firefox 96.0.2
- `cat /sys/module/nvidia_drm/parameters/modeset` gives Y
- I have configured the Firefox options as described in the README, excluding the last one which doesn't apply to my GPU driver version

I'm on a relatively fresh install (2 days old) and I haven't made any manual changes related to Nvidia or graphics other than following the README of this repo.

Edit: 
I have since followed these steps but the issues remain:
https://www.reddit.com/r/kde/comments/p44ilu/fps_drops_and_laggy_animations_under_kde_plasma/h8x4epd/

```zsh
$ vainfo
vainfo: VA-API version: 1.13 (libva 2.13.0)
vainfo: Driver version: VA-API NVDEC driver
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            : VAEntrypointVLD
      VAProfileMPEG2Main              : VAEntrypointVLD
      VAProfileVC1Simple              : VAEntrypointVLD
      VAProfileVC1Main                : VAEntrypointVLD
      VAProfileVC1Advanced            : VAEntrypointVLD
      <unknown profile>               : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointVLD
      VAProfileH264High               : VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointVLD
      VAProfileVP8Version0_3          : VAEntrypointVLD
```

Please let me know if I can provide more information.

**Comment by elFarto** [comment]
I know the exact thing you're seeing, as I've seen it myself. I believe what you're seeing is the frames in decode order, not playback order.
The library uses a small pool of images to export the frames to Firefox, I think what's happening is we're exporting over a frame that's still in use. I have a suspicion that this and issue #36 are related. 

Those videos you posted don't seem to have an issue on my machine, but I'll try and find one that does and see if I can fix it.

**Comment by Munzu** [comment]
Thanks for the confirmation! That does sound plausible, when I first observed it, my first thought was that the frames are time traveling but I couldn't say it for sure. 

> Those videos you posted don't seem to have an issue on my machine

Have you tried it with the h264ify plugin? Maybe you're not seeing it because your GPU can decode VP9 and and you'll be able to reproduce it by forcing YouTube to use AVC1 with the plugin.

**Comment by elFarto** [comment]
> Maybe the pool is too small? Another issue might be the need for explicit sync. When writing the nvdec/vulkan interop in mpv, I had to use semaphores to synchronise the reads and writes of the shared surfaces to avoid this problem. GL has a fuzzier sync model so I'm not sure what you'd have to do here.

There are a couple of things that are potentially causing an issue. Firstly the EGLStream we're using to transfer the frame from CUDA to our EGL instance is currently in mailbox mode, which means it's likely only using 2 EGLImages. I'll post an update later changing that to FIFO mode, as soon as I can find a good video to reproduce the issue.

Second is that we're reusing EGLImages that we're exporting. I don't think there's an extra copy in there, so it means we can change the image after we export the frame. It might be worth destroying the images afterwards, but that seems inefficient.

**Comment by elFarto** [comment]
Can you try the latest git version? I've added more logging to try and track down the issue. The more I thought about it, the less I liked the idea that the EGLStream was to blame, so I think collecting some more detailed logs is a better course of action. You can try out the FIFO mode by swapping those lines out in export-buf.c.

**Comment by Munzu** [comment]
If that's directed at me, I'd like to help but I don't really know what to do. Could you walk me through it? I've only ever used the AUR package. There is an [AUR git version](https://aur.archlinux.org/packages/nvidia-vaapi-driver-git/) of the package but I don't know how to apply changes to its code.

Should I build it from source? If so, is this what I have to do?

* Uninstall AUR package
* Clone the repo
* Installing dependencies for building
* `cd` into repo
* `meson setup build`
* `meson install -C build`
* Reboot

Anything else? 
If I do it this way, how do I uninstall it again? 

> You can try out the FIFO mode by swapping those lines out in export-buf.c.

Which lines exactly do I need to switch? [Line 111](https://github.com/elFarto/nvidia-vaapi-driver/blob/master/src/export-buf.c#L111) and [Line 112](https://github.com/elFarto/nvidia-vaapi-driver/blob/master/src/export-buf.c#L111)?

Sorry about the nooby questions, like I said I've never dealt with something like this before.

**Comment by elFarto** [comment]
Ah, I didn't realise you were using the AUR package. You got the steps correct, except you don't need to reboot, just close all instances of Firefox.

Don't worry about the FIFO testing, the logs should be more than enough to show what's going on.

**Comment by Munzu** [comment]
I see.

> Don't worry about the FIFO testing, the logs should be more than enough to show what's going on.

In that case just installing the git version of the package should work right? Since it basically installs from source for me.

https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nvidia-vaapi-driver-git

**Comment by elFarto** [comment]
I've not used Arch before, but if installing that package builds the source from git, then you should be good to just reinstall it. The instructions in the README are the abbreviated version, they do all the steps in the PKGBUILD.

**Comment by Munzu** [comment]
Thank you. I installed the git version package and restarted Firefox with the following command:
```zsh
LIBVA_DRIVER_NAME=nvidia MOZ_DISABLE_RDD_SANDBOX=1 NVD_LOG=1 firefox
```

Here's the output. I hope I did everything right.
https://gist.github.com/Munzu/ed8eea34f932333967406ca9581e30a7

**Comment by elFarto** [comment]
Yep, you did everything correct! Thank you. However, I'm not sure the log file actually shows anything useful, which is some what concerning. I had expected to see some mismatched items. There's really only one line in there that seems out of place, and it's not really that unusual.

I'll keep looking into it.

**Comment by Munzu** [comment]
Thank you for your hard work. Let me know if there's anything I can do.

**Comment by NoXPhasma** [comment]
I've had those jitter on some pages as well with H264 videos, it was fixed for me by disabling Vsync for Firefox. You can try if it helps you as well. For that add `__GL_SYNC_TO_VBLANK=0` to the environment variables for Firefox: `__GL_SYNC_TO_VBLANK=0 LIBVA_DRIVER_NAME=nvidia MOZ_DISABLE_RDD_SANDBOX=1 firefox`

**Comment by Munzu** [comment]
> `__GL_SYNC_TO_VBLANK=0 LIBVA_DRIVER_NAME=nvidia MOZ_DISABLE_RDD_SANDBOX=1 firefox`

Thanks but I tried that out and it didn't work for me unfortunately.

Also, I think I already had that in my Nvidia settings

![image](https://user-images.githubusercontent.com/31477160/150881000-814375a1-d95b-4a16-aea2-133acf4c95a1.png)

**Comment by Munzu** [comment]
I found another good test video, it seems this one was specifically designed to test the codec and framerate:

https://drive.google.com/file/d/0BwxFVkl63-lETTVtaVV0LUVRZ1k/view?resourcekey=0-rQVL39j5O_oysqA4xOSefQ

From time to time, I can see the bars slightly teleporting and not sliding smoothly.

I found it on this site which lists a bunch of test footage of all kinds of codecs and formats:
https://kodi.wiki/view/Samples#Codecs.2C_Framerates.2C_Black-levels_and_Subtitles

This video is the entry called "8. H.264 1080p/60 (MKV)" 

I also tested ["9. VC-1 1080p/23.976"](https://drive.google.com/file/d/0BwxFVkl63-lETFVpOWRVVGNfeEU/view?resourcekey=0-e9HrVO61IGkoTc_GBMNbww) since my GPU should be supporting it too according to the [Nvidia website](https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new#geforce). I don't think I'm seeing any jitters with this one but it's harder to say because of the low framerate. When I show the "Stats for nerds" they both show up with the same codec though (`avc1.42001E, mp4a.40.2 (22)`) so maybe the google drive player converts them to that format.

**Comment by elFarto** [comment]
I have found a video that can occasionally reproduce it for me[1], if I play it back at 1.25x speed. I have also managed to capture video of it happening, so I was able to step through frame-by-frame to see that it's display an old frame rather than a new one. This likely means it's a synchronisation issue, but I'm somewhat confused as to why it's only happening on certain computers/videos.

[1] https://www.twitch.tv/videos/1263550592, skip to the ~44 minute mark.

**Comment by Munzu** [comment]
I can confirm I'm seeing the same kind of jitters with your VOD, but only at 1.25x speed.

**Comment by elFarto** [comment]
So good news, I think I understand the problem, although I'm only vaguely sure I know why it only occurs on certain videos. What you're seeing with the jittering is a newer frame being overwritten by an older one. This happens as I believe the encoding sequence is different to the display sequence, so the frames maybe displayed ABCD but they're encoded ADBC. In this case D is clobbered by B's frame. Or something like that. Again I only vaguely understand what's going on here.

The bad news is that I need to make large changes to the export mechanism. This isn't unexpected, the export function has been largely unchanged since I managed to get it working a few months ago. The root cause of the issue (and #36) is that it's using just 2 textures to export frames, and I need to increase that to as many as the user needs.

This change was already pretty likely due to how Chrome seems to want to support VA-API.

**Comment by philipl** [comment]
Yeah, B frames mean you will decode a frame as a dependency of another frame that will be presented before it. In general, your  stream will require a sufficiently large decoded picture buffer, which is your surface pool. I guess that the subtlety here, which we don't have in mpv, for example, is that the vaapi wrapper is not a media client, and it is not handling reordering, so it is still passing frames out in decode order. In contrast, by the time the hwdec interop code in mpv is mapping frames, they are already in presentation order.

The end result is that the wrapper has to preserve the DPB semantics, and hold onto frames until they are ready for presentation.

**Comment by elFarto** [comment]
What makes this worse is that NVDEC has its own pool of decode surfaces, and now I'll have a set of decode surfaces, and they all have to go through an output surface. And I don't even know how many decode surfaces we need. Plus Chrome looks like it allocates then exports all the surfaces up-front, not the continuous method that ffmpeg uses.

That makes this whole thing take far more memory than is necessary.

**Comment by philipl** [comment]
I have not looked at the ownership semantics closely, but is it possible to create an EGLImage and then export it and release all references on our side and have the consumer hold the last ref? The caller of vaExportSurfaceHandle is required to close the fd, etc. Can we get rid of the EGLImage and still have the exported fd be valid? If so, you can just create a new one each time, pass it out, and then wash your hands of it.

**Comment by elFarto** [comment]
In theory yes, that should work, it just doesn't get you anything. You still need to duplicate a surface to export for each decode surface. Also for Chrome it looks like it exports once and just keeps the texture mapped and assumes any vaEndPicture will update that texture directly, so you won't get the option to do that.

I guess we'll just have to see how it goes. The code for this isn't looking too bad. We allocate a pair of CUarray's and push it into EGL the first time the client exports that surface. Then any other time they call export on it, we can just export the existing EGLImage directly. I might even try exporting it when we create it, then dup'ing the fds for each Export call.

**Comment by elFarto** [comment]
@Munzu I'm not sure how easy it is for you to compile from a branch, but I've push some new code to the export-rewrite branch that should hopefully help with the issue you're seeing.

**Comment by joebonrichie** [comment]
Can confirm that fixes the frame pacing jittering I was experiencing playing YT at 2x speed

(good reference video as you can focus on smooth pedalling and itś obvious if something is amiss (https://youtu.be/oqgZ5g8KzFI?t=213)

**Comment by mohad12211** [comment]
Wow! the new branch fixed this issue for me and #36 as well.

**Comment by Munzu** [comment]
I can confirm it does seem to fix this issue and #36 for me as well!
I'm observing some rare and very tiny stutters but I suspect they're just ordinary frame drops since I can see them with software decoding as well (by turning off the h264ify plugin).

This is insane, thank you @elFarto and everyone involved for getting this issue fixed so quickly!

**Comment by elFarto** [comment]
Excellent, and thanks for testing. I'll get this merged in later.

Issue #37 is open for the minor stuttering that we're seeing, it may just be a Firefox issue.
